httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/GGSSchemeBase.java

```diff
@@ -60,69 +60,167 @@
+                }
+                state = State.TOKEN_READY;
+                break;
+            case TOKEN_SENT:
+                if (challenged) {
+                    state = State.TOKEN_READY;
+                } else if (mutualAuth){
+                    // We should have received a valid mutualAuth token
+                    if (!gssContext.isEstablished()) {
+                        if (LOG.isDebugEnabled()) {
+                            final HttpClientContext clientContext =
+                                    HttpClientContext.adapt(context);
+                            final String exchangeId = clientContext.getExchangeId();
+                            LOG.debug("{} GSSContext is not established ", exchangeId);
+                        }
+                        state = State.FAILED;
+                        // TODO should we have specific exception(s) for these ?
+                        throw new AuthenticationException(
+                                "requireMutualAuth is set but GSSContext is not established");
+                    } else if (!gssContext.getMutualAuthState()) {
+                        if (LOG.isDebugEnabled()) {
+                            final HttpClientContext clientContext =
+                                    HttpClientContext.adapt(context);
+                            final String exchangeId = clientContext.getExchangeId();
+                            LOG.debug("{} requireMutualAuth is set but GSSAUthContext does not have"
+                                    + " mutualAuthState set", exchangeId);
+                        }
+                        state = State.FAILED;
+                        throw new AuthenticationException(
+                                "requireMutualAuth is set but GSSContext mutualAuthState is not set");
+                    } else {
+                        state = State.SUCCEEDED;
+                    }
+                }
+                break;
+            default:
+                state = State.FAILED;
+                throw new IllegalStateException("Illegal state: " + state);
+
+            }
+        } catch (final GSSException gsse) {
+            state = State.FAILED;
+            if (gsse.getMajor() == GSSException.DEFECTIVE_CREDENTIAL
+                    || gsse.getMajor() == GSSException.CREDENTIALS_EXPIRED) {
+                throw new InvalidCredentialsException(gsse.getMessage(), gsse);
+            }
+            if (gsse.getMajor() == GSSException.NO_CRED) {
+                throw new InvalidCredentialsException(gsse.getMessage(), gsse);
+            }
+            if (gsse.getMajor() == GSSException.DEFECTIVE_TOKEN
+                    || gsse.getMajor() == GSSException.DUPLICATE_TOKEN
+                    || gsse.getMajor() == GSSException.OLD_TOKEN) {
+                throw new AuthenticationException(gsse.getMessage(), gsse);
+            }
+            // other error
+            throw new AuthenticationException(gsse.getMessage(), gsse);
```

Comment:
Won't those duplicate the message?

Reply:
Not really.
But the Auth. execption would have the same message as the cause.
We can write some new message for the Auth. exception, I just really hate when a library swallows the original exception instead of wrapping it.